Sources/MapboxMaps/Foundation/Extensions/Core/MapEvents.swift

```diff
@@ -68,274 +68,98 @@ import Foundation
-                self = .styleLoaded
-            case MapEvents.styleImageMissing:
-                self = .styleImageMissing
-            case MapEvents.styleImageRemoveUnused:
-                self = .styleImageRemoveUnused
-            case MapEvents.sourceDataLoaded:
-                self = .sourceDataLoaded
-            case MapEvents.sourceAdded:
-                self = .sourceAdded
-            case MapEvents.sourceRemoved:
-                self = .sourceRemoved
-            case MapEvents.renderFrameStarted:
-                self = .renderFrameStarted
-            case MapEvents.renderFrameFinished:
-                self = .renderFrameFinished
-            case MapEvents.cameraChanged:
-                self = .cameraChanged
-            case MapEvents.resourceRequest:
-                self = .resourceRequest
-            default:
-                return nil
-            }
-        }
-
-        public var rawValue: String {
-            switch self {
-            case .mapLoaded:
-                return MapEvents.mapLoaded
-            case .mapLoadingError:
-                return MapEvents.mapLoadingError
-            case .mapIdle:
-                return MapEvents.mapIdle
-            case .styleDataLoaded:
-                return MapEvents.styleDataLoaded
-            case .styleLoaded:
-                return MapEvents.styleLoaded
-            case .styleImageMissing:
-                return MapEvents.styleImageMissing
-            case .styleImageRemoveUnused:
-                return MapEvents.styleImageRemoveUnused
-            case .sourceDataLoaded:
-                return MapEvents.sourceDataLoaded
-            case .sourceAdded:
-                return MapEvents.sourceAdded
-            case .sourceRemoved:
-                return MapEvents.sourceRemoved
-            case .renderFrameStarted:
-                return MapEvents.renderFrameStarted
-            case .renderFrameFinished:
-                return MapEvents.renderFrameFinished
-            case .cameraChanged:
-                return MapEvents.cameraChanged
-            case .resourceRequest:
-                return MapEvents.resourceRequest
-            }
+    /// Typed events emitted by the SDK
+    struct Event<Payload: Decodable> {
+        /// The style has been fully loaded, and the map has rendered all visible tiles.
+        public static var mapLoaded: Event<NoPayload> { .init(name: MapEvents.mapLoaded) }
+
+        /// Describes an error that has occurred while loading the Map. The `type` property defines what resource could
+        /// not be loaded and the `message` property will contain a descriptive error message.
+        /// In case of `source` or `tile` loading errors, `source-id` will contain the id of the source failing.
+        /// In case of `tile` loading errors, `tile-id` will contain the id of the tile.
+        public static var mapLoadingError: Event<MapLoadingErrorPayload> { .init(name: MapEvents.mapLoadingError) }
```

Comment:
As we already reimplementing event types, should we consider unionization of these events? 
It's now possible to have a single event with `Result<Void, MapLoadingErrorPayload>` in payload type, right?

Reply:
Interesting idea, I guess we could, indeed, have a single event, however it would be inconsistent with other events and a bit messy to implement(this combined event maps to two `MapsEvents`).

Reply:
One of my points was that migration to the async/await functions would be more native if these events were combined.
However, I just understand that the function `onNext(event:)` should be `throws` by default as it is not a thing for most of the events.
Probably, we have to keep the layout of the existing events until we find some better way to utilize that.

Reply:
I also assume that `Result` does not confirm `Decodable` so we cannot use it as a payload type.